src/Tasks/Copy.cs

```diff
@@ -374,15 +429,137 @@ CopyFileWithState copyFile
+                }
+                sourceIndices.Add(i);
+            }
+
+            var successFlags = new IntPtr[DestinationFiles.Length];  // Lockless flags updated from each thread.
+            var actionBlockOptions = new ExecutionDataflowBlockOptions { MaxDegreeOfParallelism = parallelism };
+            var partitionCopyActionBlock = new ActionBlock<List<int>>(
+                async (List<int> partition) =>
+                {
+                    // Break from synchronous thread context of caller to get onto thread pool thread.
+                    await System.Threading.Tasks.Task.Yield();
+
+                    for (int partitionIndex = 0; partitionIndex < partition.Count; partitionIndex++)
+                    {
+                        int fileIndex = partition[partitionIndex];
+                        ITaskItem sourceItem = SourceFiles[fileIndex];
+                        ITaskItem destItem = DestinationFiles[fileIndex];
+                        string sourcePath = sourceItem.ItemSpec;
+
+                        // Check if we just copied from this location to the destination, don't copy again.
+                        bool copyComplete = partitionIndex > 0 &&
+                                            String.Equals(
+                                                sourcePath,
+                                                SourceFiles[partition[partitionIndex - 1]].ItemSpec,
+                                                StringComparison.OrdinalIgnoreCase);
+
+                        if (!copyComplete)
+                        {
+                            if (DoCopyIfNecessary(
+                                new FileState(sourceItem.ItemSpec),
+                                new FileState(destItem.ItemSpec),
+                                copyFile))
+                            {
+                                copyComplete = true;
+                            }
+                            else
+                            {
+                                // Thread race to set outer variable but they race to set the same (false) value.
+                                success = false;
+                            }
+                        }
+
+                        if (copyComplete)
+                        {
+                            sourceItem.CopyMetadataTo(destItem);
+                            successFlags[fileIndex] = (IntPtr)1;
+                        }
+                    }
+                },
+                actionBlockOptions);
+
+            foreach (List<int> partition in partitionsByDestination.Values)
+            {
+                bool partitionAccepted = partitionCopyActionBlock.Post(partition);
+                if (!partitionAccepted)
+                {
+                    // Retail assert...
+                    throw new InvalidOperationException(
+                        "BUGCHECK: Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity");
```

Comment:
To more closely match our style elsewhere, I'd say

```c#
ErrorUtilities.VerifyThrow(partitionAccepted,
    "Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
```

That throws an `InternalErrorException` which is handled semi-specially.

Reply:
fixed